contracts/adapters/CharmAdapter.sol

```diff
@@ -0,0 +1,378 @@
+        this.populateOTokenMappings();
+      }
+
+      tokenAddress = lookupOToken(optionTerms);
+
+      require(tokenAddress != address(0), "Market needs to exist!");
+
+      // Get strike index, whether is long
+      AdapterStorageTypes.CharmOptionType memory optionType = adapterStorage.addressToOptionType(tokenAddress);
+      //get token
+      IOptionToken token = IOptionToken(tokenAddress);
+      //get market
+      IOptionMarket market = IOptionMarket(token.market());
+      IERC20 baseToken = IERC20(optionTerms.optionType == ProtocolAdapterTypes.OptionType.Put ? optionTerms.strikeAsset : optionTerms.underlying);
+
+      uint256 amountIn;
+
+      if(baseToken.isETH()){
+        amountIn = market.buy{value: address(this).balance}(
+          optionType.isLongToken,
+          optionType.strikeIndex,
+          amount,
+          maxCost
+        );
+      }else{
+        uint256 premium = optionViews.getBuyOptionCost(market, optionType.isLongToken, optionType.strikeIndex, amount);
+
+        uint256 balanceBefore = baseToken.uniBalanceOf(address(this));
+        baseToken.uniTransferFromSenderToThis(premium);
+        uint256 balanceAfter = baseToken.uniBalanceOf(address(this));
+
+        baseToken.safeApprove(address(market), balanceAfter.sub(balanceBefore));
+
+        amountIn = market.buy(
+          optionType.isLongToken,
+          optionType.strikeIndex,
+          amount,
+          maxCost
+        );
+      }
+
+      emit Purchased(
+          msg.sender,
+          _name,
+          optionTerms.underlying,
+          amountIn,
+          0
+      );
+    }
+
+    /**
+     * @notice Exercises the options contract.
+     * @param options is the address of the options contract
+     * @param amount is the amount of tokens or options contract to exercise. Only relevant for fungle protocols like Opyn
+     * @param recipient is the account that receives the exercised profits. This is needed since the adapter holds all the positions and the msg.sender is an instrument contract.
+     */
+    function exercise(
+        address options,
+        uint256,
+        uint256 amount,
+        address recipient
+    ) public payable override {
+
+      // Get strike index, whether is long
+      AdapterStorageTypes.CharmOptionType memory optionType = adapterStorage.addressToOptionType(options);
+
+      //get token
+      IOptionToken token = IOptionToken(options);
+      //get market
+      IOptionMarket market = IOptionMarket(token.market());
+
+      // if we are exercising but market has not settled, do it
+      if(market.isExpired() && !market.isSettled()){
+        market.settle();
+      }
+
+      uint256 profit = exerciseProfit(options, 0, amount);
+      require(profit > 0, "Not profitable to exercise");
+
+      uint256 amountOut = market.sell(optionType.isLongToken, optionType.strikeIndex, amount, profit);
+
+      //transfer over
+      market.baseToken().uniTransfer(payable(recipient), amountOut);
```

Comment:
if the option is exercised and it is a put, the transfer will be in USDC, not the underlying. So there will be no exchange to the underlying in this case. wanted to make sure you noticed it and are ok with this